File: web/src/main/java/no/nav/modiapersonoversikt/legacy/sak/service/SakOgBehandlingService.java
package no.nav.modiapersonoversikt.legacy.sak.service;

import no.nav.modiapersonoversikt.service.pdl.PdlOppslagService;
import no.nav.modiapersonoversikt.legacy.sak.providerdomain.Behandling;
import no.nav.modiapersonoversikt.legacy.sak.providerdomain.FeilendeBaksystemException;
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.binding.SakOgBehandlingV1;
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.informasjon.finnsakogbehandlingskjedeliste.Behandlingskjede;
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.informasjon.finnsakogbehandlingskjedeliste.Sak;
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.meldinger.FinnSakOgBehandlingskjedeListeRequest;
import org.slf4j.Logger;

import org.springframework.beans.factory.annotation.Autowired;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;
import static no.nav.modiapersonoversikt.legacy.sak.providerdomain.Baksystem.SAK_OG_BEHANDLING;
import static no.nav.modiapersonoversikt.legacy.sak.service.filter.Filter.filtrerBehandlinger;
import static no.nav.modiapersonoversikt.legacy.sak.service.filter.Filter.filtrerSaker;
import static no.nav.modiapersonoversikt.legacy.sak.transformers.Transformers.TIL_BEHANDLING;
import static org.slf4j.LoggerFactory.getLogger;

public class SakOgBehandlingService {

    private static final Logger logger = getLogger(SakOgBehandlingService.class);

    @Autowired
    private SakOgBehandlingV1 sakOgBehandlingPortType;

    @Autowired
    private PdlOppslagService pdlOppslagService;

    private static List<Behandling> hentBehandlingerFraBehandlingskjeder(List<Behandlingskjede> behandlingskjedeListe) {
        return behandlingskjedeListe.stream()
                .map(TIL_BEHANDLING)
                .collect(toList());
    }

    protected List<Sak> hentAlleSaker(String fnr) {
        try {
            String aktorId = pdlOppslagService.hentAktorId(fnr);
            FinnSakOgBehandlingskjedeListeRequest request = new FinnSakOgBehandlingskjedeListeRequest();
            request.setAktoerREF(aktorId);
            List<Sak> sobSaker = sakOgBehandlingPortType.finnSakOgBehandlingskjedeListe(request).getSak();
            return filtrerSaker(sobSaker);
        } catch (RuntimeException ex) {
            logger.error("Det skjedde en uventet feil mot Sak og Behandling", ex);
            throw new FeilendeBaksystemException(SAK_OG_BEHANDLING);
        }
    }

    public Map<String, List<no.nav.modiapersonoversikt.legacy.sak.providerdomain.Behandlingskjede>> hentBehandlingskjederGruppertPaaTema(String fnr) {
        return hentAlleSaker(fnr)
                .stream()
                .collect(toMap(SAKSTEMA, (wsSak) -> tilBehandligskjeder(wsSak)));
    }

    private List<Behandling> filtrerteBehandlinger(Sak sak) {
        return filtrerBehandlinger(hentBehandlingerFraBehandlingskjeder(sak.getBehandlingskjede()));
    }

    private static final Function<Sak, String> SAKSTEMA = wsSak -> wsSak.getSakstema().getValue();

    private static final Function<Behandling, no.nav.modiapersonoversikt.legacy.sak.providerdomain.Behandlingskjede> TIL_BEHANDLINGSKJEDE = behandling -> new no.nav.modiapersonoversikt.legacy.sak.providerdomain.Behandlingskjede()
            .withStatus(behandling.getBehandlingsStatus())
            .withSistOppdatert(LocalDateTime.from(behandling.getBehandlingDato().toGregorianCalendar().toZonedDateTime()));

    private List<no.nav.modiapersonoversikt.legacy.sak.providerdomain.Behandlingskjede> tilBehandligskjeder(Sak wsSak) {
        return filtrerteBehandlinger(wsSak)
                .stream()
                .map(TIL_BEHANDLINGSKJEDE)
                .collect(toList());
    }
}

File: web/src/main/java/no/nav/modiapersonoversikt/consumer/sakogbehandling/SakOgBehandlingService.kt
package no.nav.modiapersonoversikt.consumer.sakogbehandling

import no.nav.common.types.identer.AktorId
import no.nav.common.types.identer.EksternBrukerId
import no.nav.common.types.identer.Fnr
import no.nav.modiapersonoversikt.service.pdl.PdlOppslagService
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.binding.SakOgBehandlingV1
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.informasjon.finnsakogbehandlingskjedeliste.Sak
import no.nav.tjeneste.virksomhet.sakogbehandling.v1.meldinger.FinnSakOgBehandlingskjedeListeRequest
import org.slf4j.LoggerFactory
import java.time.LocalDateTime
import java.time.ZoneId

interface SakOgBehandlingService {
    fun hentAlleSaker(brukerId: EksternBrukerId): List<Sak>
    fun hentBehandlingskjederGruppertPaTema(brukerId: EksternBrukerId): Map<String, Behandlingskjede>

    class Behandling(
        val tema: String,
        val behandlingsType: String,
        val behandlingsStatus: BehandlingsStatus,
        val behandlingsDato: LocalDateTime,
        val sisteBehandlingRef: String
    ) {
        val prefix: String by lazy { sisteBehandlingRef.substring(0, 2) }
    }

    enum class BehandlingsStatus {
        UNDER_BEHANDLING, FERDIG_BEHANDLET, AVBRUTT
    }

    class Behandlingskjede(
        val tema: String,
        val status: BehandlingsStatus,
        val sistOppdatert: LocalDateTime,
    )
}

class SakOgBehandlingServiceImpl(
    val sakOgBehandlingV1: SakOgBehandlingV1,
    val pdl: PdlOppslagService
) : SakOgBehandlingService {
    companion object {
        val ulovligPrefix = "17"
        val ulovligeSakstema = listOf("FEI", "SAK", "SAP", "OPP", "YRA", "GEN", "AAR", "KLA", "HEL")
        val lovligeBehandlingstyper = listOf("ae0047", "ae0034", "ae0014", "ae0020", "ae0019", "ae0011", "ae0045")
        val log = LoggerFactory.getLogger("SakOgBehandlingServiceImpl")
    }

    override fun hentAlleSaker(brukerId: EksternBrukerId): List<Sak> {
        val aktorId = when (brukerId) {
            is AktorId -> brukerId.get()
            is Fnr -> pdl.hentAktorId(brukerId.get())
            else -> error("Kan bare bruke Fnr/AktorId man fikk ${brukerId::class.simpleName}")
        }
        val response = sakOgBehandlingV1.finnSakOgBehandlingskjedeListe(
            FinnSakOgBehandlingskjedeListeRequest().apply {
                aktoerREF = aktorId
            }
        )

        return response.sak
            .asSequence()
            .filtrerSaker() // Tilsvarerer gamle hentAlleSaker
            .toList()
    }

    override fun hentBehandlingskjederGruppertPaTema(brukerId: EksternBrukerId): Map<String, List<SakOgBehandlingService.Behandlingskjede>> {
        val cutoff = LocalDateTime.now().minusMonths(1)
        return hentAlleSaker(brukerId)
            .asSequence()
            .flatMap { sak -> hentBehandlingerFraKjede(sak) }
            .filter { it.behandlingsType in lovligeBehandlingstyper }
            .filter { it.behandlingsStatus != SakOgBehandlingService.BehandlingsStatus.AVBRUTT }
            .filter { it.prefix != ulovligPrefix }
            .filter {
                if (it.behandlingsStatus == SakOgBehandlingService.BehandlingsStatus.FERDIG_BEHANDLET) {
                    it.behandlingsDato.isAfter(cutoff)
                } else {
                    true
                }
            }
            .sortedByDescending { it.behandlingsDato }
            .map {
                SakOgBehandlingService.Behandlingskjede(
                    tema = it.tema,
                    status = it.behandlingsStatus,
                    sistOppdatert = it.behandlingsDato
                )
            }
    }

    private fun hentBehandlingerFraKjede(sak: Sak): List<SakOgBehandlingService.Behandling> {
        return sak.behandlingskjede.map {
            val behandlingsStatus = when (val status = it.sisteBehandlingsstatus.value) {
                "opprettet" -> SakOgBehandlingService.BehandlingsStatus.UNDER_BEHANDLING
                "avbrutt" -> SakOgBehandlingService.BehandlingsStatus.AVBRUTT
                "avsluttet" -> SakOgBehandlingService.BehandlingsStatus.FERDIG_BEHANDLET
                else -> error("Ukjent behandlingsstatus mottatt: $status")
            }

            val dato = if (behandlingsStatus == SakOgBehandlingService.BehandlingsStatus.FERDIG_BEHANDLET) {
                if (it.slutt == null) {
                    log.warn(
                        """
                        Inkonsistent data fra sak og behandling: Behandling rapporteres som avsluttet uten at kjede har slutt-tid satt.
                        Bruker `sisteBehandlingsoppdatering` som fallback.
                        BehandlingsId: ${it.sisteBehandlingREF}
                        BehandlinsgkjedeId: ${it.behandlingskjedeId}
                        """.trimIndent()
                    )
                }
                it.slutt ?: it.sisteBehandlingsoppdatering
            } else {
                it.start
            }.toGregorianCalendar().time

            SakOgBehandlingService.Behandling(
                tema = sak.sakstema.value,
                sisteBehandlingRef = it.sisteBehandlingREF,
                behandlingsType = it.sisteBehandlingstype.value,
                behandlingsStatus = behandlingsStatus,
                behandlingsDato = LocalDateTime.ofInstant(dato.toInstant(), ZoneId.systemDefault()),
            )
        }
    }

    private fun Sequence<Sak>.filtrerSaker(): Sequence<Sak> {
        return this
            .filter(::harLovligSakstema)
            .filter(::harBehandlinger)
            .filter(::harMinstEnLovligBehandling)
    }

    private fun Sequence<SakOgBehandlingService.Behandling>.filtrerBehandlinger(): Sequence<SakOgBehandlingService.Behandling> {
        val cutoff = LocalDateTime.now().minusMonths(1)
        return this
            .filter(::harLovligBehandlingstype)
            .filter(::harLovligBehandlingsstatus)
            .filter(::harLovligPrefix)
            .filter { sisteBehandlingErIkkeELdreEnn1Maned(cutoff, it) }
            .sortedBy { it.behandlingsDato }
    }

    private fun Sequence<SakOgBehandlingService.Behandling>.tilBehandlingskjeder(): Sequence<SakOgBehandlingService.Behandlingskjede> {
        return this
            .map {
                SakOgBehandlingService.Behandlingskjede(
                    tema = it.tema,
                    status = it.behandlingsStatus,
                    sistOppdatert = it.behandlingsDato
                )
            }
    }

    private fun harLovligSakstema(sak: Sak): Boolean {
        TODO()
    }

    private fun harBehandlinger(sak: Sak): Boolean {
        TODO()
    }

    private fun harMinstEnLovligBehandling(sak: Sak): Boolean {
        TODO()
    }

    private fun harLovligBehandlingstype(behandling: SakOgBehandlingService.Behandling): Boolean {
        return behandling.behandlingsType in lovligeBehandlingstyper
    }

    private fun harLovligBehandlingsstatus(behandling: SakOgBehandlingService.Behandling): Boolean {
        return behandling.behandlingsStatus != SakOgBehandlingService.BehandlingsStatus.AVBRUTT
    }

    private fun harLovligPrefix(behandling: SakOgBehandlingService.Behandling): Boolean {
        return behandling.prefix != ulovligPrefix
    }

    private fun sisteBehandlingErIkkeELdreEnn1Maned(
        cutoff: LocalDateTime,
        behandling: SakOgBehandlingService.Behandling
    ): Boolean {
        return if (behandling.behandlingsStatus == SakOgBehandlingService.BehandlingsStatus.FERDIG_BEHANDLET) {
            behandling.behandlingsDato.isAfter(cutoff)
        } else {
            true
        }
    }
}
